opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -1303,24 +1304,9 @@ namespace Opm
     typename MultisegmentWell<TypeTag>::EvalWell
     MultisegmentWell<TypeTag>::
     getSegmentRateUpwinding(const int seg,
-                            const int comp_idx,
-                            const bool upwinding,
-                            int& seg_upwind) const
+                            const int comp_idx) const
     {
-        // not considering upwinding for the injectors for now
-        if ((!upwinding) || (well_type_ == INJECTOR) || (primary_variables_evaluation_[seg][GTotal] <= 0.) || (seg == 0)) {
```

Comment:
The only explanation is `primary_variables_evaluation_` and `primary_variables_` do not have the same values under some circumstances.

Reply:
There should be an `initPrimaryVariablesEvaluation()` after https://github.com/OPM/opm-simulators/blob/c554fdb5fb016406b5540c69def2fe474e349d12/opm/simulators/wells/BlackoilWellModel_impl.hpp#L948 .

Reply:
Basically, in the master branch, when solveWellEq failed, it tries to recover the old primary variables. But it does not recover the evaluation version of the primary variables, which are the ones that work when assembling the well equations, so the recovering is not very complete. And some inconsistency is introduced there. 

In the function `assemble`, 
https://github.com/OPM/opm-simulators/blob/c554fdb5fb016406b5540c69def2fe474e349d12/opm/simulators/wells/BlackoilWellModel_impl.hpp#L668

When solveWellEq does not get converged for the first time, maybe we should not enter the second solveWellEq, since the well_state is still the old one, not improved by the first solveWellEq at all.